python/tank/util/shotgun/publish_resolve.py

```diff
@@ -0,0 +1,469 @@
+    # C:\Documents and Settings\davris\FileSchemeURIs.doc -> file:///C:/Documents%20and%20Settings/davris/FileSchemeURIs.doc
+    #
+    # >>> urlparse.urlparse("file:///C:/Documents%20and%20Settings/davris/FileSchemeURIs.doc")
+    # ParseResult(scheme='file', netloc='', path='/C:/Documents%20and%20Settings/davris/FileSchemeURIs.doc', params='', query='', fragment='')
+    #
+    # for information about windows, see
+    # https://blogs.msdn.microsoft.com/ie/2006/12/06/file-uris-in-windows/
+
+    if parsed_url.netloc:
+        # unc path
+        resolved_path = urllib.unquote("//%s%s" % (parsed_url.netloc, parsed_url.path))
+    else:
+        resolved_path = urllib.unquote(parsed_url.path)
+
+    # python returns drive letter paths incorrectly and need adjusting.
+    if re.match("^/[A-Za-z]:/", resolved_path):
+        resolved_path = resolved_path[1:]
+
+    # we now have one of the following three forms (with slashes):
+    # /path/to/file.ext
+    # d:/path/to/file.ext
+    # //share/path/to/file.ext
+    log.debug("Path extracted from url: '%s'" % resolved_path)
+
+    # create a lookup table of shotgun paths,
+    # keyed by upper case storage name
+    log.debug("Building cross-platform path resolution lookup table:")
+    storage_lookup = {}
+    for storage in get_cached_local_storages(tk):
+        storage_key = storage["code"].upper()
+        storage_lookup[storage_key] = ShotgunPath.from_shotgun_dict(storage)
+        log.debug("Added Shotgun Storage %s: %s" % (storage_key, storage_lookup[storage_key]))
+
+    # get default environment variable set
+    # note that this may generate a None/None/None entry
+    storage_lookup["_DEFAULT_ENV_VAR_OVERRIDE"] = ShotgunPath(
+            os.environ.get("SHOTGUN_PATH_WINDOWS"),
+            os.environ.get("SHOTGUN_PATH_LINUX"),
+            os.environ.get("SHOTGUN_PATH_MAC")
+        )
+    log.debug("Added default env override: %s" % storage_lookup["_DEFAULT_ENV_VAR_OVERRIDE"])
+
+    # look for storage overrides
+    for env_var in os.environ.keys():
+        expr = re.match("^SHOTGUN_PATH_(WINDOWS|MAC|LINUX)_(.*)$", env_var)
+        if expr:
+            platform = expr.group(1)
+            storage_name = expr.group(2).upper()
+            log.debug(
+                "Added %s environment override for %s: %s" % (platform, storage_name, os.environ[env_var])
+            )
+
+            if storage_name not in storage_lookup:
+                # not in the lookup yet. Add it
+                storage_lookup[storage_name] = ShotgunPath()
+
+            if platform == "WINDOWS":
+                storage_lookup[storage_name].windows = os.environ[env_var]
+            elif platform == "MAC":
+                storage_lookup[storage_name].macosx = os.environ[env_var]
+            else:
+                storage_lookup[storage_name].linux = os.environ[env_var]
+
+    # now see if the given url starts with any storage def in our setup
+    for storage, sg_path in storage_lookup.iteritems():
```

Comment:
Since `storage_lookup` is a throw away dictionary that is iterated on until we find a match, maybe writing a generator would be cleaner and more reusable?